### PR TITLE
daemon/stats: more resilient cpu sampling

### PR DESCRIPTION
**Pull request by stevvooe**
To avoid noise in sampling CPU usage metrics, we now sample the system
usage closer to the actual response from the underlying runtime. Because
the response from the runtime may be delayed, this makes the sampling
more resilient in loaded conditions. In addition to this, we also
replace the tick with a sleep to avoid situations where ticks can backup
under loaded conditions.

The trade off here is slightly more load reading the system CPU usage
for each container. There may be an optimization required for large
amounts of containers but the cost is on the order of 15 ms per 1000
containers. If this becomes a problem, we can time slot the sampling,
but the complexity may not be worth it unless we can test further.

Unfortunately, there aren't really any good tests for this condition.
Triggering this behavior is highly system dependent. As a matter of
course, we should qualify the fix with the users that are affected.

Signed-off-by: Stephen J Day <stephen.day@docker.com>

Short Nosed Bear

### CODE REVIEW

File: daemon/stats/collector.go [hunk 1]
```diff
@@ -106,6 +108,8 @@ func (s *Collector) Run() {
 				logrus.Errorf("collecting stats for %s: %v", pair.container.ID, err)
 			}
 		}
+
+		time.Sleep(s.interval)
```
Comment (ripcurld) [inline daemon/stats/collector.go]: Why did you use `time.Sleep` and not kept using `range time.Tick`?
Unless I am wrong, they are similar in functionality.

Comment (stevvooe) [reply daemon/stats/collector.go]: If it takes more than the interval time, the ticks backup and cause a burst of sampling, making the issue worse.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by crosbymichael** [comment]
LGTM

**Comment by ripcurld** [comment]
LGTM :trophy:

**Review by vdemeester — APPROVED** [review]
LGTM :tiger2:

**Pull request merged**